### PR TITLE
Add feedback API

### PR DESCRIPTION
**Pull request by palango**
Part of #336 

This adds the feedback API to the PFS, I'd like to wait with merging this until I have the client changes done and tested.

Basically this PR adds a new ` /<token_network_address>/feedback` endpoint, which can be used to send feedback about routes to the PFS.

Some points for discussion:
- I don't store the sender of the feedback, as we can't be sure the person requesting a route is the actual source in the request. We'd need a signature in the request for that.
- I decided to return HTTP 200, even if the feedback was not accepted, the reasons for this are that the client's doesn't need to know and
- The time interval for feedback is currently set to 10mins
- Feedback is not yet persisted to the DB

### CODE REVIEW

File: src/pathfinding_service/api.py [hunk 1]
```diff
@@ -152,7 +155,22 @@ def post(self, token_network_address: str) -> Tuple[dict, int]:
...
             )
 
-        return {"result": paths}, 200
+        # Create a feedback token and store it to the DB
+        feedback_token = create_and_store_feedback_token(
+            pathfinding_service=self.pathfinding_service,
+            token_network_address=token_network.address,
+        )
+
+        return {"result": {"paths": paths, "feedback_token": feedback_token.id.hex}}, 200
```
Comment (palango) [inline src/pathfinding_service/api.py]: We could undo this change and merge the PR, and apply this change as soon as the client changes are ready.

Comment (karlb) [reply src/pathfinding_service/api.py]: If we have one token for multiple paths, we'll have to be able to store different results for different token+path combinations. That's ok, it was just something I didn't think about before.

Comment (karlb) [reply src/pathfinding_service/api.py]: We could also keep all keys on the top level and have `{"result": ..., "token": ...}`. Then we would save one level of nesting and be able to merge it without breaking compatibility.

File: src/pathfinding_service/api.py [hunk 2]
```diff
@@ -253,6 +271,54 @@ def get(
...
+        feedback_request = self._parse_post(FeedbackRequest)
+        feedback_token = self.pathfinding_service.database.get_feedback_token(
+            token_id=feedback_request.token, token_network_address=token_network.address
+        )
+
+        # The client doesn't need to know whether the feedback was accepted or not,
+        # so in case the token is invalid we still return HTTP 200
+        if not feedback_token:
+            return {}, 200
+        if not feedback_token.is_valid():
```
Comment (karlb) [inline src/pathfinding_service/api.py]: Since both of these will do the same thing, you could fold them into a single if `if not feedback_token or not feedback_token.is_valid()`

File: src/pathfinding_service/api.py [hunk 3]
```diff
@@ -253,6 +271,54 @@ def get(
...
+            return {}, 200
+
+        log.info("Received feedback", feedback=feedback_request)
+        token_network.feedback.append(
+            RouteFeedback(
+                status=feedback_request.status,
+                received_time=datetime.utcnow(),
+                path=feedback_request.path,
+            )
```
Comment (karlb) [inline src/pathfinding_service/api.py]: We should take care that only one response per token+route will be stored. But we can do that when we add persistence.

File: src/pathfinding_service/model/feedback.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    def is_valid(self) -> bool:
+        """ Checks if the token is valid."""
+        return self.creation_time + MAX_AGE_OF_FEEDBACK_REQUESTS > datetime.utcnow()
+
+
+@dataclass
+class RouteFeedback:
+    status: str
+    received_time: datetime
+    path: List[Address]
```
Comment (karlb) [inline src/pathfinding_service/model/feedback.py]: When we want to deduplicate token+path combinations, we'll have to add the token, here.

File: src/pathfinding_service/schema.sql [hunk 5]
```diff
@@ -52,5 +52,6 @@ CREATE TABLE capacity_update (
 CREATE TABLE feedback_token (
     token_id CHAR(32) NOT NULL,
     creation_time TIMESTAMP NOT NULL,
+    token_network_address CHAR(42) NOT NULL,
```
Comment (karlb) [inline src/pathfinding_service/schema.sql]: We won't be able to make sure that the client only provides feedback for the routes we've sent to him unless we save those routes. But we can do that at a later time.

File: tests/pathfinding/test_api.py [hunk 6]
```diff
@@ -328,3 +324,58 @@ def make_params(timestamp: datetime):
...
+    assert response.json()["error_code"] == exceptions.InvalidRequest.error_code
+
+    # Test valid IOU, but not in PFS DB
+    token = FeedbackToken(token_network_address=token_network_model.address)
+
+    response = make_request(token_id=token.id.hex)
+    assert response.status_code == 200
+    assert len(token_network_model.feedback) == 0
+
+    # Test old IOU
```
Comment (karlb) [inline tests/pathfinding/test_api.py]: IOU?

File: tests/pathfinding/test_api.py [hunk 7]
```diff
@@ -328,3 +324,58 @@ def make_params(timestamp: datetime):
...
+    response = make_request(status="abc")
+    assert response.status_code == 400
+    assert response.json()["error_code"] == exceptions.InvalidRequest.error_code
+
+    # Request with invalid path
+    response = make_request(path="abc")  # type: ignore
+    assert response.status_code == 400
+    assert response.json()["error_code"] == exceptions.InvalidRequest.error_code
+
+    # Test valid IOU, but not in PFS DB
```
Comment (karlb) [inline tests/pathfinding/test_api.py]: IOU?

**Comment by palango** [comment]
Fixed the easy review comments.

For the other ones, this PR (or a follow-up) should also:
- Create one token for each returned path
- Store the path with the token in the DB
- When feedback is received, store the feedback with a reference to the original token

@karlb Anything to add?

**Comment by karlb** [comment]
> For the other ones, this PR (or a follow-up) should also:
>     * Create one token for each returned path

This is one solution, but not necessary. The other way would be use token+route as primary key. You could even combine the token table with the result table and have the following columns

* token
* token_created_at
* route
* successful (true, false or null)
* transfer_duration
PRIMARY KEY (token, route)

The first three columns would be set during insertion when the routes are sent to the client, the last two will be updated when the response is received. This fits nicely with the requirements:
* When no entry for this token+route combination is present, you can't update the row to include the results. Which is what we want
* If the results are already set, you'll overwrite them instead of adding a second feedback entry. Which is also what we want.

File: src/pathfinding_service/api.py [hunk 8]
```diff
@@ -297,7 +302,9 @@ def post(
         token_network = self._validate_token_network_argument(token_network_address)
         feedback_request = self._parse_post(FeedbackRequest)
         feedback_token = self.pathfinding_service.database.get_feedback_token(
```
Comment (karlb) [inline src/pathfinding_service/api.py]: Instead of doing
* select
* check if feedback is present
* update

you could do all in one step:
```
UPDATE feedback
SET successful = ?, feedback_time = ? 
WHERE token_id = ? AND route = ? AND successful IS NULL
```
The results `rowcount` attribute will tell you how many rows have been changed.

File: src/pathfinding_service/database.py [hunk 9]
```diff
@@ -206,31 +208,84 @@ def get_token_networks(self) -> Iterator[TokenNetwork]:
...
+                token_id, creation_time, token_network_address, route
             ) VALUES (
                 :token_id,
                 :creation_time,
-                :token_network_address
+                :token_network_address,
+                :route
             )
         """,
             token_dict,
```
Comment (karlb) [inline src/pathfinding_service/database.py]: Not for this PR, but it's getting time that we create an insert helper like
```python
database.insert('feedback', token_dict)
```

File: src/pathfinding_service/database.py [hunk 10]
```diff
@@ -206,31 +208,84 @@ def get_token_networks(self) -> Iterator[TokenNetwork]:
...
+            """,
+            [
+                token.id.hex,
+                to_checksum_address(token.token_network_address),
+                json.dumps(hexed_route),
+            ],
+        ).fetchone()
+
+        if feedback:
+            return feedback["successful"] is not None
```
Comment (karlb) [inline src/pathfinding_service/database.py]: You can also do `SELECT successful IS NOT NULL` right in the query. To me that is more straighforward, but I'm also used to have logic in SQL.

File: src/pathfinding_service/schema.sql [hunk 11]
```diff
@@ -49,9 +49,12 @@ CREATE TABLE capacity_update (
...
 
-CREATE TABLE feedback_token (
+CREATE TABLE feedback (
     token_id CHAR(32) NOT NULL,
     creation_time TIMESTAMP NOT NULL,
     token_network_address CHAR(42) NOT NULL,
-    PRIMARY KEY (token_id)
+    route TEXT NOT NULL,
+    successful BOOLEAN CHECK (successful IN (0,1)),
+    feedback_time TIMESTAMP,
```
Comment (karlb) [inline src/pathfinding_service/schema.sql]: Is this the time at which we received the feedback or the time it took to do the transfer? It sounds like the former, but IMO the latter would be more useful. Or do we assume `feedback_time - creation_time = transfer_duration`?

File: src/pathfinding_service/api.py [hunk 12]
```diff
@@ -156,19 +156,24 @@ def post(self, token_network_address: str) -> Tuple[dict, int]:
...
+def create_and_store_feedback_tokens(
+    pathfinding_service: PathfindingService,
+    token_network_address: TokenNetworkAddress,
+    routes: List[Dict],
 ) -> FeedbackToken:
     feedback_token = FeedbackToken(token_network_address=token_network_address)
-    pathfinding_service.database.insert_feedback_token(feedback_token)
+
+    for route in routes:
+        pathfinding_service.database.prepare_feedback(token=feedback_token, route=route["path"])
```
[the remark this answers is not in the document]
Comment (palango) [reply src/pathfinding_service/api.py]: added a TODO

File: src/pathfinding_service/schema.sql (hunk 11 above)
Comment (palango) [reply src/pathfinding_service/schema.sql]: This is the time the feedback is received, my thinking was that it's better to save the raw information, the duration is then easy to calculate from them.

File: src/pathfinding_service/api.py (hunk 8 above)
Comment (palango) [reply src/pathfinding_service/api.py]: That's a nice improvement!

File: src/pathfinding_service/schema.sql (hunk 11 above)
Comment (karlb) [reply src/pathfinding_service/schema.sql]: If we can easily calculate the duration, I agree. But it might get complicated if the first request does not succeed. The client might even wait for locks to expire before trying another route. But maybe I'm too pessimistic here and deriving the data will be fine.

Comment (palango) [reply src/pathfinding_service/schema.sql]: So you say we should take the latest feedback for a given token as the start, and the current time as the end?

Comment (karlb) [reply src/pathfinding_service/schema.sql]: I think this is the only way to get timings out of this data. This will only work if the client tries each route immediately after the previous one, submits feedback immediately and the latency to the PFS is small compared to the duration of a payment.

That's why I thought making the client submit the duration explicitly would make data more robust and easier to handle.
